src/mlir/design.md

```diff
@@ -3,69 +3,225 @@
+## Arithmetic and Builtins
 
-For example, heap allocation, which LLVM IR generally uses calls to `@llvm.malloc`, Verona uses `snmalloc`, which has a finer control over what we can do with the regions.
-In the end, however, it will be the same: calls to the runtime library in IR.
+Arithmetic and builtins are implemented in Verona via direct calls to intrinsics and MLIR/LLMV operations by name (string).
+Each numeric class will have their operations defined in the standard library and the compiler will detect the call as _internal_ and create the appropriate intrinsic / operation.
+Builtins (number conversion, checks, cpu-specific instructions) will also be implemented in the standard Verona library.
 
-The generator knows what runtime calls are appropriate for each AST construct, and can insert those calls in the right places and make sure to declare the methods so that linking with the runtime library works.
+Adding two numbers (`a + b`) is the same as calling the method `+` on `a` over `b` (ex. `a.+(b)` or `+(a, b)`).
+These calls will be matched against the already implemented methods in those classes and behave like a normal call (dynamic or static).
 
-## Arithmetic
+These methods are small enough that the LLVM inliner pass will convert all calls into single operations.
+If any of them proves more complex than the inliner can handle, we can add the _always_inline_ attribute to them.
 
-Arithmetic is generated by the generator via pattern matching on the types and selectors.
-Adding two numbers (`a + b`) is the same as calling the method `+` on `a` over `b` (`a.+(b)`), but these methods don't exist (how would we implement addition if not with a `+`?).
+## Type Constraints
 
-So the generator has to find those that weren't declared and replace with arithmetic if the types are compatible (int, float).
+Eventually, all types will be clear on the AST, but this isn't the case right now.
+For now, we have to add truncations and extensions where it isn't clear.
 
-Users are still allowed to use `+` as a method name, in which case the generator will recognise and call that method instead of transforming into arithmetic, even if the types are compatible.
 For example:
+ * Constants of type `int` or `float` are created as 64-bit numbers and then truncated if assigned to 32-bit values.
+ * Arithmetic between types of different width have the smallest type promoted to the largest.
 
-```
-// Verona code
-class U32 {}
-class Foo {
-  static +(a: U32, b: U32) { ... }
-}
-main() : U32 {
-  let a = 21 + 21; // This is simple arithmetic
-  let b = Foo::+(21, 21); // This calls the method
-}
-```
+Also, due to a constraint on the standard dialect in MLIR, we currently only support sign-less integers (neither signed nor unsigned).
+Arithmetic will be lowered in the LLVM dialect, so will contain full support for signed and unsigned as well as all intrisics.
```

Comment:
I'm not quite sure I understand what this means.  LLVM also only has signless integers, it has signed or unsigned operations on integers.  Does MLIR not have these?